design/loadbalancer-hash-policy-design.md

```diff
@@ -0,0 +1,188 @@
+Depending on the `RequestAttribute`, some fields of this generic map may be required, an if missing, the hash policy excluded.
+For example, to implement header hashing, the `headerName` field would be required.
+The `Terminal` field denotes if the attribute specified (e.g. header name) is found, Envoy should stop processing any subsequent hash policies in the list it is provided.
+This is a performance optimization and can provide a speedup in hash calculation if set and the attribute to hash on is found.
+An example of how this field may be used from the Envoy docs is [here](https://www.envoyproxy.io/docs/envoy/v1.17.0/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-routeaction-hashpolicy-terminal).
+
+```
+type RequestHashPolicy struct {
+	RequestAttribute string            `json:"requestAttribute,omitempty"`
+	HashOptions      map[string]string `json:"hashOptions,omitempty"`
+	Terminal         bool              `json:"terminal,omitempty"`
+}
+```
+
+### HTTPProxy YAML Example
+
+An example of how this feature would be used to hash on a specific header value follows below:
+
+```
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+   name: example
+spec:
+  virtualhost:
+    fqdn: example.projectcontour.io
+  routes:
+  - services:
+    - name: example-app
+      port: 80
+    loadBalancerPolicy:
+      strategy: RequestHash
+      requestHashPolicies:
+      - requestAttribute: Header
+        hashOptions:
+          headerName: X-Some-Header
+        terminal: true
+      - requestAttribute: Header
+        hashOptions:
+          headerName: User-Agent
+```
+
+The value of the `X-Some-Header` header will be hashed by Envoy if present and used to make a load balancing decision.
+Consistent values in this header would lead to Envoy routing to the same backend service instance.
+If the header is present, Envoy will *not* move on to attempting to hash the `User-Agent` header value.
+
+### Handling Invalid Strategy Choice
+As per the documentation of the [`LoadBalancerPolicy` field on the `ExtensionService` object](https://projectcontour.io/docs/v1.11.0/config/api/#projectcontour.io/v1alpha1.ExtensionService), the `Cookie` load balancing strategy is invalid for use with the `ExtensionService`.
+Currently, that restriction does not seem to be enforced (or on the `TCPRoute` object either, for which the `Cookie` strategy is not compatible).
+This design proposes that we start to enforce these strategy restrictions and also restrict the new `RequestHash` strategy to only valid on `Route` objects.
+If specified on `ExtensionService` or `TCPRoute` objects, it will be overridden to the default `RoundRobin` strategy and a warning will be logged.
+Overriding rather than throwing an error is a deliberate attempt to continue to allow the configured `HTTPProxy` or configured extension to function, albeit with basic load balancing enabled.
```

Comment:
Set a warning on the object telling the user this is happening?

Reply:
whoops I thought i added a sentence about that! will add

Reply:
WDYT about adding a warning on the object if there are other lower level configuration errors as well? Realized I forgot to add that as well

Reply:
See https://github.com/projectcontour/contour/pull/3268/commits/4852bc08ee0b567204dba03a7a24a6f7ccd9861e